File: erlang/gamevalue.erl
% Perfect Tic-Tac-Toe evaluator in Erlang
% Copyright © 2016 Bart Massey
% [This program is licensed under the "MIT License"]
% Please see the file COPYING in the source
% distribution of this software for license terms.

-module(gamevalue).
-export([gamevalue/2]).

% Scan for diagonal.
has_diagonal(_, _, D) when D > 2 -> true;

has_diagonal(OnMove, Board, D) ->
    board:get(D, D, Board) == OnMove and
    has_diagonal(OnMove, Board, D + 1).
    
% Scan for opposite diagonal.
has_opposite_diagonal(_, _, D) when D > 2 -> true;

has_opposite_diagonal(OnMove, Board, D) ->
    board:get(D, 2 - D, Board) == OnMove and
    has_opposite_diagonal(OnMove, Board, D + 1).

% Check this row.
has_row(_, _, _, C) when C > 2 -> true;

has_row(OnMove, Board, R, C) ->
    board:get(R, C, Board) == OnMove and
    has_row(OnMove, Board, R, C + 1).

% Scan for rows.
has_some_row(OnMove, Board, R) when R < 3 ->
    has_row(OnMove, Board, R, 0);

has_some_row(OnMove, Board, R) when R < 3 ->
    has_some_row(OnMove, Board, R + 1).

% Check this column.
has_col(_, _, R, _) when R > 2 -> true;

has_col(OnMove, Board, R, C) ->
    board:get(R, C, Board) == OnMove and
    has_row(OnMove, Board, R + 1, C).

% Scan for columns.
has_some_col(OnMove, Board, C) when C < 3 ->
    has_col(OnMove, Board, C, 0) or
    has_some_col(OnMove, Board, C + 1).

% Succeeds if there is a win on the board for
% the side on move.
has_win(Board, OnMove) ->
    has_diagonal(Board, OnMove, 0) or
    has_opposite_diagonal(Board, OnMove, 0) or
    has_some_row(Board, OnMove, 0) or
    has_some_col(Board, OnMove, 0).

% Succeeds if there is a blank on the board.
has_move(Board, R, C) when C > 2 ->
    has_move(Board, R + 1, 0);

has_move(Board, R, C) when R =< 2 ->
    board:get(R, C, Board) /= 0 and
    has_move(Board, R, C + 1).

% Returns the value of the game to the
% side on move if the game is over, or
% -2 if the game is still in progress.
gamevalue(OnMove, Board) ->
    % This is really gross. I don't see a better plan
    % in Erlang, though.
    case has_win(OnMove, Board) of
        true -> 1;
        false -> case has_win(-OnMove, Board) of true -> -1;
        false -> case has_move(Board, 0, 0) of true -> 0;
        false -> -2
    end end end.

File: php/gamevalue.php
<?php
/*
 * Copyright © 2016 Matthew Slocum
 * [This program is licensed under the "MIT License"]
 * Please see the file COPYING in the source
 * distribution of this software for license terms.
 */

/* Perfect Tic-Tac-Toe player in PHP */

/*
  Returns the value of the game to the
  side on move if the game is over, or
  -2 if the game is still in progress.
*/
function gamevalue($onmove, &$board) {
    $v;
    /* first scan for wins */
    $side = -1;
    for ($s = 0; $s < 2; $s++) {
        $side = -$side;
        $v = $side * $onmove;
        /* scan for diagonal */
        $n = 0;
        for ($d = 0; $d < 3; $d++) {
            if ($board[$d][$d] == $side) {
                $n = $n + 1;
            }
        }
        if ($n == 3) {
            return $v;
        }
        /* scan for opposite diagonal */
        $n = 0;
        for ($d = 0; $d < 3; $d++) {
            if ($board[$d][2 - $d] == $side) {
              $n = $n + 1;
            }
        }
        if ($n == 3) {
            return $v;
        }
        /* scan for rows */
        for ($r = 0; $r < 3; $r++) {
            $n = 0;
            for ($c = 0; $c < 3; $c++) {
                if ($board[$r][$c] == $side) {
                    $n = $n + 1;
                }
            }
            if ($n == 3) {
                return $v;
            }
        }
        /* scan for columns */
        for ($c = 0; $c < 3; $c++) {
            $n = 0;
            for ($r = 0; $r < 3; $r++) {
                if ($board[$r][$c] == $side) {
                    $n = $n + 1;
                }
            }
            if ($n == 3)
                return $v;
        }
    }
    /* scan for blanks */
    for ($r = 0; $r < 3; $r++) {
        for ($c = 0; $c < 3; $c++) {
            if ($board[$r][$c] == 0) {
                /* game not over */
                return -2;
            }
            /* game is a draw */
            return 0;
        }
    }
}
 ?>
